sys/net/link_layer/ng_slip/ng_slip.c

```diff
@@ -0,0 +1,293 @@
+            _bytes++;
+        }
+    }
+}
+
+static void _slip_receive(size_t bytes)
+{
+    msg_t msg;
+    ng_pktsnip_t *hdr = ng_netif_hdr_build(NULL, 0, NULL, 0);
+    ng_netreg_entry_t *sendto;
+
+    if (hdr == NULL) {
+        DEBUG("slip: no space left in packet buffer\n");
+        return;
+    }
+
+    ((ng_netif_hdr_t *)hdr->data)->if_pid = thread_getpid();
+    ng_pktsnip_t *pkt = ng_pktbuf_add(NULL, NULL, bytes, NG_NETTYPE_UNDEF);
+
+    if (pkt == NULL) {
+        DEBUG("slip: no space left in packet buffer\n");
+        ng_pktbuf_release(hdr);
+        return;
+    }
+
+    pkt->next = hdr;
+
+    if (ringbuffer_get(_in_buf, pkt->data, bytes) != bytes) {
+        DEBUG("slip: could not read %zu bytes from ringbuffer\n", bytes);
+        ng_pktbuf_release(pkt);
+        return;
+    }
+
+#ifdef MODULE_NG_IPV6
+
+    if (ng_ipv6_hdr_get_version((ng_ipv6_hdr_t *)pkt->data) == 6) {
+        pkt->type = NG_NETTYPE_IPV6;
+    }
+
+#endif
+
+    sendto = ng_netreg_lookup(pkt->type, NG_NETREG_DEMUX_CTX_ALL);
+
+    if (sendto == NULL) {
+        DEBUG("slip: unable to forward packet of type %i\n", pkt->type);
+        ng_pktbuf_release(pkt);
+    }
+
+    ng_pktbuf_hold(pkt, ng_netreg_num(pkt->type, NG_NETREG_DEMUX_CTX_ALL) - 1);
+
+    msg.type = NG_NETAPI_MSG_TYPE_RCV;
+    msg.content.ptr = (char *)pkt;
+
+    while (sendto != NULL) {
+        DEBUG("slip: sending pkt %p to PID %u\n", pkt, sendto->pid);
+        msg_send(&msg, sendto->pid);
+        sendto = ng_netreg_getnext(sendto);
+    }
+}
+
+static void _slip_send(uart_t uart, ng_pktsnip_t *pkt)
+{
+    bool errored = false;
+    ng_pktsnip_t *ptr;
+
+    ptr = pkt->next;    /* ignore ng_netif_hdr_t, we don't need it */
+
+    while ((ptr != NULL) && !errored) {
+        DEBUG("slip: send pktsnip of length %zu over UART_%d\n", ptr->size, uart);
+        uint8_t *data = ptr->data;
+
+        for (size_t i = 0; i < ptr->size; i++) {
+            switch (data[i]) {
+                case _SLIP_END:
+                    DEBUG("slip: encountered END byte on send: stuff with ESC\n");
+
+                    if (uart_write(uart, _SLIP_ESC) < 0) {
+                        errored = true;
+                        break;
+                    }
+
+                    if (uart_write(uart, _SLIP_END_ESC) < 0) {
+                        errored = true;
+                        break;
+                    }
+
+                    break;
+
+                case _SLIP_ESC:
+                    DEBUG("slip: encountered ESC byte on send: stuff with ESC\n");
+
+                    if (uart_write(uart, _SLIP_ESC) < 0) {
+                        errored = true;
+                        break;
+                    }
+
+                    if (uart_write(uart, _SLIP_ESC_ESC) < 0) {
+                        errored = true;
+                        break;
+                    }
+
+                    break;
+
+                default:
+                    if (uart_write(uart, data[i]) < 0) {
```

Comment:
this will not work (you will lose chars). Use `uart_write_blocking()` when you use the uart like this. Better (and I really would to this here): use the transmit interrupt and a transmit buffer!

```c
static ringbuffer_t *out_buf;

_slip_send(..) {
    ...
    default:
    ringbuffer_add_one(&out_buf, data[i]);
    uart_tx_start(uart);
    ...
}

// and
int _slip_rx_cb(void *arg) {
    if (_out_buf.avail > 0) {
        char c = (char)ringbuffer_get_one(&out_buf);
        uart_write(uart, c);
        return 1;
    }
    return 0;
}
```